### PR TITLE
Fixing the debugger "wrong namespace" issues. 

### PR DESCRIPTION
**Pull request by vspinu**
Fixing #420,  clojure-emacs/cider#2039 and clojure-emacs/cider#2004

### CODE REVIEW

File: src/cider/nrepl/middleware/debug.clj [hunk 1]
```diff
@@ -189,7 +190,7 @@ this map (identified by a key), and will `dissoc` it afterwards."}
   "Like `read`, but reply is sent through `debugger-message`.
   type is sent in the message as :input-type."
   [extras type prompt]
-  (let [key (u/random-uuid-str)
+  (let [key   (u/random-uuid-str)
```
Comment (expez) [inline src/cider/nrepl/middleware/debug.clj]: Think you forgot to turn off your alignment minor mode :)

**Comment by expez** [comment]
Think a note in the changelog for the next cider release would be nice too.  I know a lot of people have been tripped up by this one.

Great stuff as usual, @vspinu! :+1:

File: src/cider/nrepl/middleware/debug.clj (hunk 1 above)
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: It's not really a minor mode - just a toggle in `clojure-mode`. ;-)

File: src/cider/nrepl/middleware/debug.clj [hunk 2]
```diff
@@ -227,18 +237,19 @@ this map (identified by a key), and will `dissoc` it afterwards."}
         nil)
       (finally
         (try
-          (in-ns ns)
+          ;; Restore original ns in case it was changed by the evaled code.
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: evaled -> evaluated

File: src/cider/nrepl/middleware/debug.clj [hunk 3]
```diff
@@ -412,19 +426,22 @@ this map (identified by a key), and will `dissoc` it afterwards."}
...
-                               :line line, :column column}
-                              ;; There's an nrepl bug where the column starts counting
-                              ;; at 1 if it's after the first line. Since this is a
-                              ;; top-level sexp, a (= col 1) is much more likely to be
-                              ;; wrong than right.
-                              (update :column #(if (= % 1) 0 %))))
-                  :forms @*tmp-forms*}]
+  `(let [~'META__ {:msg ~(let [{:keys [code id file line column ns]} *msg*]
+                           (-> {:code code,
+                                ;; Passing namespace as :original-ns stalls. Bug?
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: stalls what?

File: src/cider/nrepl/middleware/debug.clj [hunk 4]
```diff
@@ -583,12 +602,7 @@ this map (identified by a key), and will `dissoc` it afterwards."}
                  (h (maybe-debug msg)))
       "debug-instrumented-defs" (instrumented-defs-reply msg)
       "debug-input" (when-let [pro (@promises (:key msg))]
-                      (swap! promises dissoc  (:key msg))
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: How do those changes fit with the rest?

File: src/cider/nrepl/middleware/debug.clj [hunk 5]
```diff
@@ -536,12 +536,14 @@ this map (identified by a key), and will `dissoc` it afterwards."}
...
+  [{:keys [code ns] :as msg}]
   ;; The best way of checking if there's a #break reader-macro in
   ;; `code` is by reading it, in which case it toggles `has-debug?`.
-  (let [has-debug? (atom false)
-        fake-reader (fn [x] (reset! has-debug? true) nil)]
-    (binding [*data-readers* (->> (repeat fake-reader)
+  (let [has-debug?  (atom false)
+        ;; Don't return nil in reader (https://dev.clojure.org/jira/browse/CLJ-1138)
+        fake-reader (fn [x] (reset! has-debug? true) x)]
+    (binding [*ns*           (find-ns (symbol (or ns "user")))
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: We don't really use alignment in bindings in `cider-nrepl`.

File: src/cider/nrepl/middleware/debug.clj [hunk 6]
```diff
@@ -207,12 +207,14 @@ this map (identified by a key), and will `dissoc` it afterwards."}
...
-  [form locals]
+  If an exception is thrown, it is caught and sent to the client, and this
+  function returns nil. `extras` is a metadata map received from `break'."
+  [form extras]
   (let [ns (ns-name *ns*)]
     (try
-      (binding [*tmp-locals* locals]
+      (binding [*tmp-locals* (:locals extras)
+                ;; evaluate in the instrumentation ns
+                *ns*         (find-ns (symbol (:original-ns extras)))]
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: Same here. What will happen if `:original-ns` is nil here?

File: src/cider/nrepl/middleware/debug.clj [hunk 7]
```diff
@@ -207,12 +207,14 @@ this map (identified by a key), and will `dissoc` it afterwards."}
 
 (defn- eval-with-locals
   "`eval` form wrapped in a let of the locals map.
-  If an exception is thrown, it is caught and sent to the client, and
-  this function returns nil."
-  [form locals]
+  If an exception is thrown, it is caught and sent to the client, and this
+  function returns nil. `extras` is a metadata map received from `break'."
+  [form extras]
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: I'd name `extras` something else - without reading its docstring it was hard for me to figure out what it is. `break-state` perhaps?

File: src/cider/nrepl/middleware/debug.clj [hunk 8]
```diff
@@ -227,18 +229,19 @@ this map (identified by a key), and will `dissoc` it afterwards."}
...
-          (in-ns ns)
+          ;; Restore original ns in case it was changed by the evaled code.
           ;; See https://github.com/clojure-emacs/cider/issues/1604
+          (in-ns ns)
           (catch IllegalStateException _))))))
 
 (defn- read-debug-eval-expression
   "Read and eval an expression from the client.
-  extras is a map to be added to the message, and prompt is added into
+  `extras` is a map received from `break`, and `prompt` is added into
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: Same here.

**Comment by bbatsov** [comment]
> Think a note in the changelog for the next cider release would be nice too. I know a lot of people have been tripped up by this one.

Yeah, once this is merged it should have a changelog entry for sure. Probably it will be the final bug-fix for 0.15.1. Seems no other major problems with 0.15 have been discovered in the weeks after its release.

File: src/cider/nrepl/middleware/debug.clj (hunk 1 above)
Comment (vspinu) [inline src/cider/nrepl/middleware/debug.clj]: What's wrong with this? You don't want aligned key-val pairs in cider projects?

File: src/cider/nrepl/middleware/debug.clj (hunk 2 above)
Comment (vspinu) [inline src/cider/nrepl/middleware/debug.clj]: Was saving few chars not to get into fill.

File: src/cider/nrepl/middleware/debug.clj (hunk 4 above)
Comment (vspinu) [inline src/cider/nrepl/middleware/debug.clj]: There is no rest, the only consumer of that promise is the `read-debug` itself.

File: src/cider/nrepl/middleware/debug.clj (hunk 5 above)
Comment (vspinu) [inline src/cider/nrepl/middleware/debug.clj]: Let bindings as well?

File: src/cider/nrepl/middleware/debug.clj (hunk 6 above)
Comment (vspinu) [inline src/cider/nrepl/middleware/debug.clj]: `:original-ns` can never be nil, it's always assigned in `with-initial-debug-bindings`.

File: src/cider/nrepl/middleware/debug.clj (hunk 7 above)
Comment (vspinu) [inline src/cider/nrepl/middleware/debug.clj]: Good sugestion. That map contains both current breakpoint state and current debug session state, but I think `break-state` is informative of both.

File: src/cider/nrepl/middleware/debug.clj (hunk 1 above)
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: Yes. I'm not fond of this style and avoid it everywhere.

File: src/cider/nrepl/middleware/debug.clj (hunk 4 above)
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: I see.

File: src/cider/nrepl/middleware/debug.clj (hunk 5 above)
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: Yes.

File: src/cider/nrepl/middleware/debug.clj (hunk 7 above)
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: Maybe `debugger-state` then with two maps inside of it?

**Comment by vspinu** [comment]
Good to go AFAIC. Addressed comments, fixed tests and added some extras:
 - improved and disentangled error handling during debug eval commands (eval, inspect, inject, locals), 
 - refactored naming conventions here and there

File: src/cider/nrepl/middleware/debug.clj [hunk 9]
```diff
@@ -185,20 +183,41 @@ this map (identified by a key), and will `dissoc` it afterwards."}
...
+                  ~val
+                  ;; Borrowed from `interruptible-eval/evaluate`.
+                  (catch Exception e#
+                    (let [root-ex# (#'clojure.main/root-cause e#)]
+                      (when-not (instance? ThreadDeath root-ex#)
+                        (debugger-send
+                          {:status :eval-error
+                           :causes [(let [causes# (stacktrace/analyze-causes e# (:pprint-fn *msg*))]
+                                      (when (coll? causes#) (last causes#)))]})))
+                    ::error))]
```
Comment (Malabarba) [inline src/cider/nrepl/middleware/debug.clj]: It would be more robust (and just as easy) to use a generated symbol here.

Comment (vspinu) [inline src/cider/nrepl/middleware/debug.clj]: Good point. Changed.

**Comment by Malabarba** [comment]
LGTM

**Comment by bbatsov** [comment]
👍

**Pull request merged**